File: quiz/src/main/java/com/example/quiz/controller/RolePermissionController.java
package com.example.quiz.controller;

import com.example.quiz.model.entity.role_permission.Permission;
import com.example.quiz.model.entity.role_permission.Role;
import com.example.quiz.model.entity.user.User;
import com.example.quiz.service.impl.RolePermissionServiceImpl;
import com.example.quiz.validators.requirePermission.ResourceController;
import lombok.RequiredArgsConstructor;
import lombok.extern.slf4j.Slf4j;
import org.springframework.web.bind.annotation.*;

import java.util.List;
import java.util.Set;

@RestController
@RequestMapping("/api/v1/role-permissions")
@ResourceController("ROLE_PERMISSION")
@RequiredArgsConstructor
@Slf4j
public class RolePermissionController {

    private final RolePermissionServiceImpl service;

    @GetMapping("/roles")
    public List<Role> getAllRoles() {
        return service.getAllRoles();
    }

    @PostMapping("/roles")
    public Role createRole(@RequestParam String authority, @RequestParam(required = false) String description) {
        return service.createRole(authority, description);
    }

    @PutMapping("/roles/{id}")
    public Role updateRole(@PathVariable Long id, @RequestParam String authority, @RequestParam(required = false) String description) {
        return service.updateRole(id, authority, description);
    }

    @DeleteMapping("/roles/{id}")
    public void deleteRole(@PathVariable Long id) {
        service.deleteRole(id);
    }

    @GetMapping("/permissions")
    public List<Permission> getAllPermissions() {
        return service.getAllPermissions();
    }

    @PostMapping("/permissions")
    public Permission createPermission(@RequestParam String name, @RequestParam String description, @RequestParam String resource, @RequestParam String action) {
        return service.createPermission(name, description, resource, action);
    }

    @PutMapping("/permissions/{id}")
    public Permission updatePermission(@PathVariable Long id, @RequestParam String name, @RequestParam String description, @RequestParam String resource, @RequestParam String action) {
        return service.updatePermission(id, name, description, resource, action);
    }

    @DeleteMapping("/permissions/{id}")
    public void deletePermission(@PathVariable Long id) {
        service.deletePermission(id);
    }

    @PostMapping("/roles/{roleId}/permissions")
    public void assignPermissionsToRole(@PathVariable Long roleId, @RequestBody Set<Long> permissionIds) {
        service.assignPermissionsToRole(roleId, permissionIds);
    }

    @PostMapping("/roles/{roleId}/users/{userId}")
    public void assignRoleToUser(@PathVariable Long userId, @PathVariable Long roleId) {
        service.assignRoleToUser(userId, roleId);
    }

    @PostMapping("/roles/{roleId}/users")
    public void assignUsersToRole(@PathVariable Long roleId, @RequestBody Set<Long> userIds) {
        // reuse service.assignRoleToUser for each user
        if (userIds == null || userIds.isEmpty()) return;
        userIds.forEach(userId -> service.assignRoleToUser(userId, roleId));
    }

    @GetMapping("/roles/{authority}/users")
    public List<User> getUsersByRole(@PathVariable String authority) {
        return service.getUsersByRole(authority);
    }
}

File: quiz-be/src/role-permissions/role-permission.controller.ts
import {
  Body,
  Controller,
  Delete,
  Get,
  Param,
  ParseIntPipe,
  Post,
  Put,
  Query,
  UseGuards,
} from '@nestjs/common';
import { Roles } from '../auth/decorators/roles.decorator';
import { JwtAuthGuard } from '../auth/guards/jwt-auth.guard';
import { RolesGuard } from '../auth/guards/roles.guard';
import { ResponseMessage } from '../common/decorators/response-message.decorator';
import { RolePermissionService } from './role-permission.service';

@Controller('role-permissions')
@UseGuards(JwtAuthGuard, RolesGuard)
@Roles('ADMIN', 'SUPER_ADMIN')
export class RolePermissionController {
  constructor(private readonly service: RolePermissionService) {}

  @Get('roles')
  @ResponseMessage('Success')
  async getAllRoles() {
    return this.service.getAllRoles();
  }

  @Post('roles')
  @ResponseMessage('Success')
  async createRole(
    @Query('authority') authority: string,
    @Query('description') description?: string,
  ) {
    return this.service.createRole(authority, description);
  }

  @Put('roles/:id')
  @ResponseMessage('Success')
  async updateRole(
    @Param('id', ParseIntPipe) id: number,
    @Query('authority') authority: string,
    @Query('description') description?: string,
  ) {
    return this.service.updateRole(id, authority, description);
  }

  @Delete('roles/:id')
  @ResponseMessage('Success')
  async deleteRole(@Param('id', ParseIntPipe) id: number) {
    await this.service.deleteRole(id);
    return null;
  }

  @Get('permissions')
  @ResponseMessage('Success')
  async getAllPermissions() {
    return this.service.getAllPermissions();
  }

  @Post('permissions')
  @ResponseMessage('Success')
  async createPermission(
    @Query('name') name: string,
    @Query('description') description: string,
    @Query('resource') resource: string,
    @Query('action') action: string,
  ) {
    return this.service.createPermission(name, description, resource, action);
  }

  @Put('permissions/:id')
  @ResponseMessage('Success')
  async updatePermission(
    @Param('id', ParseIntPipe) id: number,
    @Query('name') name: string,
    @Query('description') description: string,
    @Query('resource') resource: string,
    @Query('action') action: string,
  ) {
    return this.service.updatePermission(
      id,
      name,
      description,
      resource,
      action,
    );
  }

  @Delete('permissions/:id')
  @ResponseMessage('Success')
  async deletePermission(@Param('id', ParseIntPipe) id: number) {
    await this.service.deletePermission(id);
    return null;
  }

  @Post('roles/:roleId/permissions')
  @ResponseMessage('Success')
  async assignPermissionsToRole(
    @Param('roleId', ParseIntPipe) roleId: number,
    @Body() permissionIds: number[],
  ) {
    return this.service.assignPermissionsToRole(roleId, permissionIds || []);
  }

  @Post('roles/:roleId/users/:userId')
  @ResponseMessage('Success')
  async assignRoleToUser(
    @Param('roleId', ParseIntPipe) roleId: number,
    @Param('userId') userId: string,
  ) {
    await this.service.assignRoleToUser(roleId, userId);
    return null;
  }

  @Post('roles/:roleId/users')
  @ResponseMessage('Success')
  async assignUsersToRole(
    @Param('roleId', ParseIntPipe) roleId: number,
    @Body() userIds: Array<string | number>,
  ) {
    await this.service.assignUsersToRole(
      roleId,
      (userIds || []).map((id) => String(id)),
    );
    return null;
  }

  @Get('roles/:authority/users')
  @ResponseMessage('Success')
  async getUsersByRole(@Param('authority') authority: string) {
    return this.service.getUsersByRole(authority);
  }
}
